### DISCUSSION TITLE
Universal functions not everywehere?

### DISCUSSION THREAD

**Discussion by elbertov** [dbuezas/lovelace-plotly-graph-card#212] [Q&A] [answered]
Hi,

You wrote:

_Work in progress.
Functions are allowed everywhere with this format:_

```yaml
type: custom:plotly-graph-dev
entities:
  - entity: sensor.garden_temperature
    offset: $fn () => "0d"
    name: |
      $fn({ys, vars}) => {
        vars.title = ys[0];
        return ys[ys.length-1]
      }
  - entity: sensor.keller_humidity
title: $fn({ vars }) => vars.title
hours_to_show: 2
```

I try to plot a graph for the current _calenderday_ so from midnight to midnight:
- hours_to_show: 24
- time_offset: xx           this depends on the current time: at 6am this must be 18, at 20:00 this must be 4

for instance:

![image](https://user-images.githubusercontent.com/58105716/215541330-0a8159b8-d993-4967-949d-729a747c48b1.png)

So it has to be calculated, but unfortunately a $fn for time_offset doesn't work:

![image](https://user-images.githubusercontent.com/58105716/215541391-3b2a9e87-46cb-49fa-b790-3e1ecc7e66d3.png)

```yaml
type: custom:plotly-graph
entities:
  - entity: sensor.buienradar_temperature
    time_offset: $fn () => "1d"
    name: |
      $fn({vars}) => {
        vars.my_title = "Temperature outside";

        var now = new Date()
        var tomorrow = new Date(now)
        tomorrow.setDate(tomorrow.getDate() + 1)
        tomorrow.setHours(0,0,0,0)
        var HoursLeftToday = (tomorrow.getTime() - now.getTime()) / 1000 / 3600;  // milliseconds -> hours
        
        vars.my_time_offset = HoursLeftToday.toFixed(1) + "h";                    // add "h" to the duration
        console.log(vars.my_time_offset);
        
        return "yesterday";
      }
  - entity: sensor.buienradar_temperature
    name: todayy
title: $fn({ vars }) => vars.my_title
hours_to_show: 24
#time_offset: $fn({ vars }) => vars.my_time_offset
time_offset: 6.3h
```

Why does it work for the title and not for time_offset?

regards martin

**Comment by dbuezas** [comment, marked as answer]
## TL;DR: The error message explains it 

Functions inside entities need to run after the data is fetched, and to know what to fetch, the time_offset needs to be known beforehand. But the time_offsets hasn't ran yet because it's written below the entities.
When you write plain values instead of functions, this order is not enforced because it is known that the value won't change and there won't be any side effects (like storing in vars)

So pretty much this:

<img width="500" alt="image" src="https://user-images.githubusercontent.com/777196/215555244-b3e72f1e-677c-4d66-aa7c-36f192189937.png">

## Solution

Luckily, the computation you do for the entity's name doesn't depend on the fetched data, so you can extract it and put it above entities:
```yaml
not_a_prop: $fn ({ vars }) => {.
  ...
  vars.my_title = "...";
  vars.my_time_offset = "...";
}
time_offset: $fn({ vars }) => vars.my_time_offset # before entities as the error message indicates
entities: 
   - entity: ..
     name: $fn ({ ... }) => ...
```

This way, everything is computed before it is needed without requiring 1.21 gigawatts

**Reply by elbertov** [reply]
yes, that makes sense, and it works! thx!

![image](https://user-images.githubusercontent.com/58105716/215561822-33837931-08ca-49e5-865f-8c38f6c16efe.png)

```yaml
type: custom:plotly-graph
not_a_prop: $fn ({vars}) => { 
    now = new Date();
    tomorrow = new Date(now);
    tomorrow.setDate(tomorrow.getDate() + 1);
    tomorrow.setHours(0,0,0,0);
    HoursLeftToday = (tomorrow.getTime() - now.getTime()) / 1000 / 3600;
    vars.my_time_offset = HoursLeftToday.toFixed(1) + "h";
  }
time_offset: $fn({ vars }) => vars.my_time_offset
entities:
  - entity: sensor.buienradar_temperature
    time_offset: 1d
    name: yesterday
  - entity: sensor.buienradar_temperature
    name: today
title: Outside temperature
hours_to_show: 24
```

**Reply by dbuezas** [reply]
This is actually a very cool idea. [Show&Tell](https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/new?category=show-and-tell) ? :)

**Reply by dbuezas** [reply]
BTW, now that I look at it, you can get rid of not_a_prop and put the function inside the time_offset directly. Alternatively, put a fn in hours_to_show instead

**Reply by elbertov** [reply]
Ah, that's cool. I tried that before, but got errormessages and I didn't understand them exactly  ;-)

![image](https://user-images.githubusercontent.com/58105716/215705314-af5004f9-202c-4030-a392-6e98a3afc1dd.png)

But when I move the _time_offset_ to the top, it works fine and it makes more sense to put to $fn inside the time_offset directly

![image](https://user-images.githubusercontent.com/58105716/215705891-674c52c3-61f8-404a-991f-dbc7d9d387a6.png)

I will post it in Show&Tell.
thx so much for your help, you are doing a great job!

maybe a new flag :-) :
```yaml
timeline_to_show: current_day # current_week, current_month, current_year
```

**Reply by dbuezas** [reply]
That's a good idea. Maybe hours_to_show can take the strings you propose, and time durations (like 1d, 1y, etc)

**Reply by dbuezas** [reply]
And then it should change the global time_offset too, so that the full day is visible
